File: 2616 (Minimize the Maximum Difference of Pairs)/java/Solution.java
class Solution {

    boolean possible(int[] nums, int p, int mid){
        int cnt = 0, i  =1;
        while(i<nums.length){
            if (Math.abs(nums[i]-nums[i-1]) <= mid){
                cnt++;
                i++;
            }
            i++;
        }
        return cnt>=p;
    }    
    public int minimizeMax(int[] nums, int p) {
        Arrays.sort(nums);
        int low = 0, high = nums[nums.length-1], ans = -1;

        while (low<=high){
            int mid = (low+high)>>1;

            if (possible(nums, p, mid)){
                ans=  mid;
                high = mid-1;
            }else{
                low = mid+1;
            }
        }
        return ans;
    }
}

File: 2616 (Minimize the Maximum Difference of Pairs)/python/Solution.py
class Solution:
    def minimizeMax(self, nums: List[int], p: int) -> int:
        nums.sort()

        low ,high = 0,nums[-1]
        ans = -1


        def possible(mid):
            cnt , i = 0,1
            while i<len(nums):
                if abs(nums[i]-nums[i-1]) <= mid:
                    cnt+=1
                    i+=1
                
                i+=1
            return cnt>=p

        while low<=high:
            mid = (low+high)>>1

            if (possible(mid)):
                ans = mid
                high = mid-1
            else:
                low  =mid+1
            
        return ans